### PR TITLE
[RLlib] Optionally don't drop last ts in v-trace calculations (APPO and IMPALA).

### PR DESCRIPTION
**Pull request by sven1977**
Add an option to APPO/IMPALA config to not drop the last ts in vtrace calculations.
- `vtrace_drop_last_ts` (default True).

First experiments with sparse reward and reward-at-end environments indicate that not dropping the last ts adds stability to the learning behavior.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by gjoliver — COMMENTED** [review]
a lot of the diffs shou go away after rebase.

File: rllib/agents/ppo/appo_tf_policy.py [hunk 1]
```diff
@@ -166,16 +168,19 @@ def reduce_mean_valid(t):
...
                 discounts=tf.cast(
-                    ~make_time_major(tf.cast(dones, tf.bool), drop_last=True),
+                    ~make_time_major(
+                        tf.cast(dones, tf.bool), drop_last=drop_last),
                     tf.float32) * policy.config["gamma"],
-                rewards=make_time_major(rewards, drop_last=True),
-                values=values_time_major[:-1],  # drop-last=True
+                rewards=make_time_major(rewards, drop_last=drop_last),
+                values=values_time_major[:-1]
+                if drop_last else values_time_major,
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/ppo/appo_tf_policy.py:183]: Yeah, not sure. LINTer says it's ok. You mean the `if` stuff, right?

File: rllib/agents/ppo/appo_tf_policy.py [hunk 2]
```diff
@@ -166,16 +168,19 @@ def reduce_mean_valid(t):
...
                 target_policy_logits=make_time_major(
-                    unpacked_old_policy_behaviour_logits, drop_last=True),
+                    unpacked_old_policy_behaviour_logits, drop_last=drop_last),
                 actions=tf.unstack(
-                    make_time_major(loss_actions, drop_last=True), axis=2),
+                    make_time_major(loss_actions, drop_last=drop_last),
+                    axis=2),
                 discounts=tf.cast(
-                    ~make_time_major(tf.cast(dones, tf.bool), drop_last=True),
+                    ~make_time_major(
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/ppo/appo_tf_policy.py:178]: `~` means NOT.
`make_time_major` transforms a tensor of shape `[B, T, ...]` into `[T, B, ...]`.

**Comment by sven1977** [comment]
Hey @gjoliver , answered your questions and removed some unrelated functionaliy/changes. Could you take another look? Thx! :)

**Review by gjoliver — APPROVED** [review]
there seems to be a bunch of unrelated diffs, maybe rebase.
but this looks good now. thanks.

File: rllib/agents/ppo/appo_tf_policy.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (gjoliver) [reply rllib/agents/ppo/appo_tf_policy.py:183]: yeah, since it's a continuation of last line. no idea, minor comment.

**Review by gjoliver — APPROVED** [review]
